### PR TITLE
Netsim limit log requests

### PR DESCRIPTION
**Pull request by Hamms**
NetSim will now only subscribe to updates to the log table when the Log Browser is open

### CODE REVIEW

**Comment by Hamms** [comment]
@islemaster could you also take a look?

File: apps/src/netsim/NetSimTable.js [hunk 1]
```diff
@@ -111,6 +125,25 @@ var NetSimTable = module.exports = function (channel, shardID, tableName) {
...
+  this.channel_.subscribe(this.tableName_, this.channelCallback_);
+};
+
+/**
+ * Unubscribes the saved callback from events for this table on our
+ * local channel. Also clears the saved callback.
+ */
+NetSimTable.prototype.unsubscribe = function () {
+  this.channel_.unsubscribe(this.tableName_, this.channelCallback_);
+  this.cannelCallback = undefined;
```
Comment (islemaster) [inline apps/src/netsim/NetSimTable.js]: s/cannelCallback/channelCallback

**Comment by islemaster** [comment]
This looks good.  Before we ship this (though not necessarily before merging this PR) I'd like us to add simple filtering to the router log modal so that opening it from the Router Tab shows only logs for the router you are connected to.  We also need to make sure that button is available in all the levels that have the router tab at all (not just those that have the 'showLogBrowserButton' option) to provide the same information that was originally available.  That probably means splitting that level option into "showLogBrowserButtonInLobby" and "showLogBrowserButtonOnRouterTab".

**Comment by Hamms** [comment]
@islemaster how's that?

**Comment by islemaster** [comment]
Nice! I like it.  Looks good to me.  This is assigned to Phil, should probably still get his opinion too.

File: apps/i18n/netsim/az_az.json [hunk 2]
```diff
@@ -52,6 +52,8 @@
   "lobbyStatusWaitingForYou": "Waiting for you...",
   "logBrowserButton": "Log Browser",
   "logBrowserHeader": "All Router Logs",
+  "logBrowserHeader_all": "All Router Logs",
```
Comment (philbogle) [inline apps/i18n/netsim/az_az.json]: Just curious, was there an automated tool for making all of these localization changes based on the initial English one?

File: apps/src/netsim/NetSimRouterLogModal.html.ejs [hunk 3]
```diff
@@ -29,7 +29,11 @@ function sortMarkupFor(sortKey) {
 <div class="modal-dialog modal-lg">
     <div class="modal-header">
       <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
-      <h4 class="modal-title"><%= i18n.logBrowserHeader() %></h4>
+      <% if (isAllRouterMode) { %>
```
Comment (philbogle) [inline apps/src/netsim/NetSimRouterLogModal.html.ejs]: Could you add a comment explaining the "parameters" (e.g. isAllRouterMode) to this view?

File: apps/src/netsim/NetSimRouterLogModal.js [hunk 4]
```diff
@@ -149,6 +177,28 @@ NetSimRouterLogModal.prototype.onSortHeaderClick_ = function (sortKey) {
...
+NetSimRouterLogModal.prototype.setRouter = function (router) {
+  this.router_ = router;
+  this.render();
+};
+
+/**
+ * Helper method to determine whether or we are currently in
+ * "All-Router" mode, or dealing with a single router. Currently is
+ * determined exclusively by the current connection state, but may later
+ * be epanded to allow switching between the two.
```
Comment (philbogle) [inline apps/src/netsim/NetSimRouterLogModal.js]: s/epanded/expanded/

File: apps/src/netsim/NetSimTable.js [hunk 5]
```diff
@@ -50,11 +50,25 @@ var NetSimTable = module.exports = function (channel, shardID, tableName) {
   }
 
   /**
+   * @type {string}
```
Comment (philbogle) [inline apps/src/netsim/NetSimTable.js]: You can write this more concisely as 

@private {string}

File: apps/src/netsim/NetSimTable.js [hunk 6]
```diff
@@ -50,11 +50,25 @@ var NetSimTable = module.exports = function (channel, shardID, tableName) {
...
    * @private
    */
   this.channel_ = channel;
-  this.channel_.subscribe(tableName, NetSimTable.prototype.onPubSubEvent.bind(this));
+  this.subscribe();
+
+  /**
+   * The callback we most recently subscribed with, so that we can
+   * cleanly unsubscribe.
+   * @type {function}
```
Comment (philbogle) [inline apps/src/netsim/NetSimTable.js]: can you declare the parameter and returns types of the function?

File: apps/src/netsim/PubSubChannel.js [hunk 7]
```diff
@@ -69,3 +69,12 @@ PubSubChannel.PusherChannel = function (pusherApiChannel) {
 PubSubChannel.PusherChannel.prototype.subscribe = function (eventName, callback) {
   this.pusherChannel_.bind(eventName, callback);
 };
+
+/**
+ * Unsubscribe a given callback from a given event
+ * @param {string} eventName
+ * @param {function} callback
```
Comment (philbogle) [inline apps/src/netsim/PubSubChannel.js]: jsdoc function arguments and return type.

File: apps/src/netsim/PubSubService.js [hunk 8]
```diff
@@ -89,3 +89,11 @@ PubSubService.PusherService = function (applicationKey) {
 PubSubService.PusherService.prototype.subscribe = function (channelID) {
   return new PubSubChannel.PusherChannel(this.api_.subscribe(channelID));
 };
+
+/**
+ * Unsubscribefrom events on a particular channel.
```
Comment (philbogle) [inline apps/src/netsim/PubSubService.js]: space before from

**Comment by philbogle** [comment]
LGTM after a few comment nits.

File: apps/i18n/netsim/az_az.json (hunk 2 above)
Comment (Hamms) [inline apps/i18n/netsim/az_az.json]: Yup! [this doc](https://github.com/code-dot-org/code-dot-org/tree/staging/i18n#apps) gives the walkthrough

File: apps/src/netsim/NetSimRouterLogModal.html.ejs (hunk 3 above)
Comment (Hamms) [inline apps/src/netsim/NetSimRouterLogModal.html.ejs]: Happily; where should that live?

Comment (philbogle) [inline apps/src/netsim/NetSimRouterLogModal.html.ejs]: Just at the head of file. I don't think we have a formal convention for this so use your best judgement; I think of these parameters as the moral equivalent of method parameters and its helpful for the reader to know what they are.

 (There must be a way to do an ".ejs" comment vs. something that would appear in the markup; that would be preferable.)

File: apps/src/netsim/NetSimTable.js (hunk 6 above)
Comment (Hamms) [inline apps/src/netsim/NetSimTable.js]: It has neither.

**Pull request merged**